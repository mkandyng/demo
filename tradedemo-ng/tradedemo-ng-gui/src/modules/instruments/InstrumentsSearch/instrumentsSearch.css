#instrumentsSearch label {
  display: block;
  margin: 5px 0;
}

#instrumentsSearch input {
  border-radius: 5px;
  margin: 4px;
  height: 27px;
}

#instrumentsSearch code {
  padding: .2em .5em;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

#instrumentsSearch .menu {
  position: absolute;
  box-sizing: border-box;
  width: 100%;
  border: 1px solid #cccccc;
}

#instrumentsSearch {
  margin-bottom:2px;
  position: relative;
  z-index: 1;
}
