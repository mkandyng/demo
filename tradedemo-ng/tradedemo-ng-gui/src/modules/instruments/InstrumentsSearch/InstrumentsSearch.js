import React, { useState, useEffect } from "react";
import ReactAutocomplete from "react-autocomplete";
import { MAX_MARKET_FEED_INSTRUMENTS } from "../index";
import PropTypes from 'prop-types';
import "./instrumentsSearch.css"

/**
 * Component for InstrumentSearch drop down, which manage the instruments state
 */
export default function InstrumentsSearch({ instruments,
                                            marketfeedInstruments,
                                            addInstrumentToMarketfeed,
                                            fetchInstruments }) {

    const [ value, setSelectItem ] = useState("");

    useEffect(() => {
        fetchInstruments();
    }, [ fetchInstruments ]);

    return (
      <div id="instrumentsSearch">
           <form>
               <ReactAutocomplete
                   value={value}
                   onChange={event => setSelectItem(event.target.value)}
                   onSelect={value => setSelectItem(value)}
                   items={updateSearchDropDown(instruments)}
                   getItemValue={item => `${item.value}`}
                   shouldItemRender={matchStocks}
                   renderItem={renderItem}
               />
               <input type="submit" value="Add" onClick={event => addItem(event,
                                                                          value,
                                                                          instruments,
                                                                          marketfeedInstruments,
                                                                          addInstrumentToMarketfeed,
                                                                          setSelectItem)}/>
           </form>
      </div>
    )
}

function updateSearchDropDown(instruments) {
   const MAX_RECORDS = 10;
   return instruments.slice(0, MAX_RECORDS)
                     .map(instrument => ({
                         value: instrument.symbol,
                         label: "(" + instrument.symbol +  ") " +  instrument.name
                     }));
};

function addItem(event,
                 value,
                 instruments,
                 marketfeedInstruments,
                 addInstrumentToMarketfeed,
                 setSelectItem) {

    const addSelectedInstrumentToMarketfeed = () => {
        const instrument = instruments.find(inst => inst.symbol === value);
        if(instrument !== undefined) {
            if(marketfeedInstruments.length >= MAX_MARKET_FEED_INSTRUMENTS) {
                alert("Max " + MAX_MARKET_FEED_INSTRUMENTS + " instruments, please remove an one to add [" + value + "]");
            } else {
                addInstrumentToMarketfeed(instrument);
                return true;
            }
        } else {
            alert("[" + value + "] is not a valid, unable to add to marketfeed");
        }
            return false;
    };

    event.preventDefault();
    setSelectItem("");
    addSelectedInstrumentToMarketfeed();
};

function renderItem(item, highlighted) {
    return (
      <div key={item.id} style={{ backgroundColor: highlighted?"#eee":"transparent"}} >
        {item.label}
      </div>
    );
}

function matchStocks(state, value) {
      return state.label.toLowerCase().indexOf(value.toLowerCase()) !== -1;
}

InstrumentsSearch.propTypes = {
    marketfeedInstruments: PropTypes.arrayOf(PropTypes.object).isRequired,
    instruments: PropTypes.arrayOf(PropTypes.object).isRequired,
    addInstrumentToMarketfeed: PropTypes.func.isRequired,
    fetchInstruments: PropTypes.func.isRequired
};
