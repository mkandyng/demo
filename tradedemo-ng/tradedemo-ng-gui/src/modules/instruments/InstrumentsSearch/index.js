import InstrumentsSearch from "./InstrumentsSearch";

export default InstrumentsSearch;
