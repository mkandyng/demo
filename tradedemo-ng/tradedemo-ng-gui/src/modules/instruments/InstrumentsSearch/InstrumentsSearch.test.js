import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import InstrumentsSearch from "./InstrumentsSearch";

/**
 * This is unit test of InstrumentsSearch component
 * It leverage snapshot to test the presentation
 */

describe("InstrumentsSearch", () => {
  const props = { instruments: [{name:"symbol"}],
                  marketfeed: {instruments:[{symbol:"symbol"}], selected: {symbol:"symbol"}},
                  addInstrumentToMarketfeed: jest.fn(),
                  fetchInstruments: jest.fn()};

    it("should render component comparing with previous snapshot shallow markup", () => {
        // Given

        // When
        const component = shallow(<InstrumentsSearch {...props} />);

        // Then
        expect(toJson(component)).toMatchSnapshot();
    });

});
