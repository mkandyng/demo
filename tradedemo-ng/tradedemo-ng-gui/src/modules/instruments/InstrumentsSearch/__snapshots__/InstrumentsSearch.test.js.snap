// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstrumentsSearch should render component comparing with previous snapshot shallow markup 1`] = `
<div
  id="instrumentsSearch"
>
  <form>
    <Autocomplete
      autoHighlight={true}
      getItemValue={[Function]}
      inputProps={Object {}}
      isItemSelectable={[Function]}
      items={
        Array [
          Object {
            "label": "(undefined) symbol",
            "value": undefined,
          },
        ]
      }
      menuStyle={
        Object {
          "background": "rgba(255, 255, 255, 0.9)",
          "borderRadius": "3px",
          "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
          "fontSize": "90%",
          "maxHeight": "50%",
          "overflow": "auto",
          "padding": "2px 0",
          "position": "fixed",
        }
      }
      onChange={[Function]}
      onMenuVisibilityChange={[Function]}
      onSelect={[Function]}
      renderInput={[Function]}
      renderItem={[Function]}
      renderMenu={[Function]}
      selectOnBlur={false}
      shouldItemRender={[Function]}
      value=""
      wrapperProps={Object {}}
      wrapperStyle={
        Object {
          "display": "inline-block",
        }
      }
    />
    <input
      onClick={[Function]}
      type="submit"
      value="Add"
    />
  </form>
</div>
`;
